Migrate Cms page create/edit actions

zuk3975 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Migration of Improve > Design > Pages page create/edit form. `Save and preview` button in forms should be implemented in separate PR
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | Forms should work as in legacy page. Test using routes `/admin-dev/index.php/improve/design/cms-pages/new` and `/admin-dev/index.php/improve/design/cms-pages/{page_id}/edit`. :warning: Assets may need to be rebuilt.

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/12995)
<!-- Reviewable:end -->

zuk3975 committed:
Initial commit. Form type creation in progress

zuk3975 committed:
Rendered form

zuk3975 committed:
Add new cms-page action

zuk3975 committed:
Added fields validation in handler

zuk3975 committed:
Added js entry points for form

tomas862 committed:
adds cms page button

tomas862 committed:
on cancel button click adds redirect back to its parent category page

zuk3975 committed:
Edit action in progress

tomas862 committed:
changes some exception names

zuk3975 committed:
Edit action in progress

zuk3975 committed:
Editing action in progress

zuk3975 committed:
Implemented Edit action

zuk3975 committed:
Enabled js url rewrite copier in form

zuk3975 committed:
Catching Domain exception in create/edit actions

jolelievre commented on src/Adapter/CMS/Page/CommandHandler/AddCmsPageHandler.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        $cms = $this->createCmsFromCommand($command);
+
+        if (false === $cms->validateFields(false)) {
```
I think you need to `validateFieldsLang` as well

jolelievre commented on src/Adapter/CMS/Page/CommandHandler/EditCmsPageHandler.php:
```diff
@@ -0,0 +1,126 @@
+        $cms = $this->createCmsFromCommand($command);
+
+        try {
+            if (false === $cms->validateFields(false)) {
```
same here `validateFieldsLang`

jolelievre commented on src/Core/Domain/CmsPage/Exception/CmsPageException.php:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Base exception for cms page sub-domain
+ */
+class CmsPageException extends DomainException
```
I just noticed that `DomainException` simply extends `Exception`, wouldn't it be better if it extended `CoreException` ?

zuk3975 replied:
@jolelievre I guess it should, but I doubt this change belongs to this PR

jolelievre replied:
sure! it was just a remark

jolelievre commented on src/PrestaShopBundle/Controller/Admin/Improve/Design/CmsPageController.php:
```diff
@@ -694,4 +790,14 @@ private function getCmsPageCategoryErrorByExceptionTypeAndCode(CmsPageCategoryEx
+     */
+    private function getErrorMessages()
+    {
+        return [];
```
Seems like error messages are missing

jolelievre commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageType.php:
```diff
@@ -0,0 +1,279 @@
+                        'class' => 'js-copier-source-title',
+                    ],
+                    'constraints' => [
+                        //@todo: typedRegexConstraint generic_name PR #12735
```
since this PR has been merged, you can replace this constraint now

jolelievre replied:
and in the following regex constraints

jolelievre commented on src/PrestaShopBundle/Form/Admin/Improve/Design/Pages/CmsPageType.php:
```diff
@@ -0,0 +1,279 @@
+                ],
+            ])
+            ->add('friendly_url', TranslatableType::class, [
+                'error_bubbling' => false,
```
I'm wondering why some fields need `error_bubbling => false` and some others not, even though they also are `TranslatableType `

zuk3975 replied:
@jolelievre that is currently required in TranslatableType when using constraint in parent level (not in options). If error bubbling isn't set to false, that parent error is shown somewhere in wrong level and is hidden by its child or something similar. Another case when using constraints in 'options' works well with default error bubbling.

jolelievre replied:
wow that's tricky 🤪
and it's not possible to define all the constraints on the same level? either only parent or only in the options? (I guess there's a reason)
I'm wondering if this could be managed automatically by `TranslatableType` in its resolve function, something like
```php
if (!isset($options['error_bubbling']) && !empty($options['constraints']) && !empty($options['options']['constraints'])) {
    $options['error_bubbling'] = false;
}
```

I'm just proposing this because who is not aware of this particularity has great chances to forget it.

zuk3975 replied:
@jolelievre  Maybe worth creating an issue to fix that confusion in some way ? (it will definitely be lost in minutes here :D )

jolelievre replied:
Yeah of course I didn't plan to modify this in this PR ^^

jolelievre replied:
I created this issue for later https://github.com/PrestaShop/PrestaShop/issues/13107

jolelievre commented on src/PrestaShopBundle/Resources/config/routing/admin/improve/design/cms_pages.yml:
```diff
@@ -5,6 +5,20 @@ admin_cms_pages_index:
+        _controller: 'PrestaShopBundle:Admin\Improve\Design\CmsPage:create'
+        _legacy_controller: AdminCmsContent
+
+admin_cms_pages_edit:
```
careful with the indentation

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/Blocks/form.html.twig:
```diff
@@ -0,0 +1,107 @@
+        'label': 'Page content'|trans({}, 'Admin.Design.Feature'),
+      }) }}
+
+      {{ ps.form_group_row(cmsPageForm.is_indexed_for_search, {}, {
```
I didn't know we now have full macro forms, that's great! ^^
We should consider refacto all the old forms in 177

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/add.html.twig:
```diff
@@ -0,0 +1,36 @@
+  <div class="row justify-content-center">
+    <div class="col">
+      {{ include('@PrestaShop/Admin/Improve/Design/Cms/Blocks/form.html.twig', {
+        'cmsCategoryParentId': app.request.query.get('id_cms_category'),
```
couldn't it be inserted by the controller instead of getting request value in the template?

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/edit.html.twig:
```diff
@@ -0,0 +1,36 @@
+  <div class="row justify-content-center">
+    <div class="col">
+      {{ include('@PrestaShop/Admin/Improve/Design/Cms/Blocks/form.html.twig', {
+        'cmsCategoryParentId': app.request.query.get('id_cms_category'),
```
same question about request

jolelievre reviewed (changes requested):
A few feedback, and this PR needs a good rebase ^^

zuk3975 committed:
rebase fixes

zuk3975 committed:
Added fieldsLang validation

zuk3975 committed:
Fixed indent in routes, Added TypedRegex constraints in form

zuk3975 committed:
Moved some params getter from view to controller

zuk3975 committed:
built cms page form asset

zuk3975 commented:
@jolelievre addressed Your comments. CmsPageController needs refactoring to fix some merged list mistakes and update error handling. Needs to be done in separate PR ofc.

zuk3975 committed:
Added missing security annotations

zuk3975 committed:
Renewed licence in controller

marionf commented:
@zuk3975 

1. Tooltips are missing, don't know how we integrate them @TristanLDD ?

![capture d'écran_1267](https://user-images.githubusercontent.com/13449658/55244056-b2a43000-5240-11e9-9edf-ac679affd27d.png)

2. There is already prohibited characters under fields

![capture d'écran_1268](https://user-images.githubusercontent.com/13449658/55244074-bb950180-5240-11e9-93e4-d6e863758ac0.png)

3. Tiny mce is missing for page content

zuk3975 commented:
> 3. Tiny mce is missing for page content

TinyMce on translatable input is a known issue and should be fixed in another PR

TristanLDD commented:
@marionf  The tooltips should be integrated under the field, the forbidden characters are part of the tooltip, they should appear at the end of the text.

matks commented:
Summary of needed changes:
- TinyMCE: is a global issue for multiple migrated pages, will be solved in a dedicated PR during code freeze
- remove "prohibited fields" from under the field and move them into the tooltip
- enable tooltips, make them appear under the field

zuk3975 committed:
Fixed form tooltips

matks commented:
Tooltips are fixed => waiting for QA 😃

zuk3975 committed:
Added missing create button and enabled page

zuk3975 committed:
fixed wrong method in route

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/Blocks/form.html.twig:
```diff
@@ -0,0 +1,111 @@
+{{ form_start(cmsPageForm) }}
+<div class="card">
+  <div class="card-header">
+    {{ 'Page'|trans({}, 'Admin.Global') }}
```
I already have it localized in _Admin.Shopparameters.Feature_. :-)

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/Blocks/form.html.twig:
```diff
@@ -0,0 +1,111 @@
+
+      {{ ps.form_group_row(cmsPageForm.meta_keyword, {}, {
+        'label': 'Meta keywords'|trans({}, 'Admin.Global'),
+        'help': 'To add "tags" click in the field, write something, and then press "Enter."'|trans({}, 'Admin.Design.Help')
```
Oops, this string has issues with its punctuation... I would prefer _To add tags, click in the field, write something, and then press the "Enter" key._, localized in _Admin.Shopparameters.Help_. Thank you.

LouiseBonnard commented on src/PrestaShopBundle/Resources/views/Admin/Improve/Design/Cms/Blocks/form.html.twig:
```diff
@@ -0,0 +1,111 @@
+
+      {{ ps.form_group_row(cmsPageForm.friendly_url, {}, {
+        'label': 'Friendly URL'|trans({}, 'Admin.Global'),
+        'help': 'Only letters and the hyphen (-) character are allowed.'|trans({}, 'Admin.Design.Feature')
```
From a Crowdin point of view, this string is associated with the right domain. But if I look at the code, it seems that this translation key corresponds to a tooltip, am I right? If it is the case, it should fall under the _Admin.Design.Help_ domain.

matks commented:
Thank you @zuk3975 and @tomas862 for the PR
Thank you @sarahdib for the QA validation
